File: haskell/Expression.hs
module Expression (Position, Variable, Text, Expression (..), Value (..), Label) where

import Type

import Data.List
import Data.Maybe
import Data.Map (Map)
import qualified Data.Map as Map
import Data.Set (Set)
import qualified Data.Set as Set

type Position = (String, Int, Int)

type Variable = String

type Text = String

data Expression
    = EAt Position Expression
    | EVariable Variable
    | EApply Expression Expression
    | ELambda Variable Expression
    | ELet Variable Expression Expression
    | EText Text
    | ECons Expression Expression
    | ENil
    | EFor Variable Expression Expression
    | EAccess Expression Label
    | ERecord (Map Label (Expression, Bool))
    | EConcat Expression Expression
    | EChoice Expression Expression
    deriving Show

data Value
    = VLambda (Map Variable Value) Variable Expression
    | VText Text
    | VList [Value]
    | VRecord (Map Label Value)
    deriving Show

File: scala/src/templ/Expression.scala
package templ

sealed trait Expression

object Expression {
  type Variable = String
  type Position = (String, Int, Int)
  type Text = String
  type Label = String
  type Required = Boolean

  case class EAt (position: Position, expression: Expression) extends Expression
  case class EVariable (variable: Variable) extends Expression
  case class EApply (function: Expression, argument: Expression) extends Expression
  case class ELambda (variable: Variable, body: Expression) extends Expression
  case class ELet (variable: Variable, value: Expression, body: Expression) extends Expression
  case class EText (text: Text) extends Expression
  case class ECons (head: Expression, tail: Expression) extends Expression
  case class ENil extends Expression
  case class EFor (variable: Variable, list: Expression, body: Expression) extends Expression
  case class ELookup (record: Expression, label: Label) extends Expression
  case class ERecord (fields: Map[Label, (Expression, Required)]) extends Expression
  case class EConcat (left: Expression, right: Expression) extends Expression
  case class EChoice (primary: Expression, secondary: Expression) extends Expression

  def errorWithPosition(message: String, position: Position) = {
    position match {
      case (filename, line, column) =>
        message + " at line " + line + ", column " + column +
                (if(!filename.isEmpty) " in '" + filename + "'" else "")
    }
  }
}
